cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -128,55 +132,70 @@
   }
 
   @Override
-  public synchronized RuntimeInfo lookup(final RunId runId) {
-    RuntimeInfo runtimeInfo = super.lookup(runId);
-    if (runtimeInfo != null) {
-      return runtimeInfo;
+  protected RuntimeInfo createRuntimeInfo(ProgramController controller, Program program) {
+    if (controller instanceof AbstractTwillProgramController) {
+      RunId twillRunId = ((AbstractTwillProgramController) controller).getTwillRunId();
+      return new SimpleRuntimeInfo(controller, program, twillRunId);
     }
+    return null;
+  }
 
-    // Lookup all live applications and look for the one that matches runId
-    String appName = null;
-    TwillController controller = null;
-    for (TwillRunner.LiveInfo liveInfo : twillRunner.lookupLive()) {
-      for (TwillController c : liveInfo.getControllers()) {
-        if (c.getRunId().equals(runId)) {
-          appName = liveInfo.getApplicationName();
-          controller = c;
-          break;
-        }
-      }
-      if (controller != null) {
-        break;
+  private synchronized boolean isTwillRunIdCached(RunId twillRunId) {
```

Comment:
Add a TODO to have a better data structure to support this more efficiently. It can be as simple as a object that contains a Table and a Set.